server/src/main/java/org/apache/iotdb/db/qp/strategy/optimizer/ConcatPathOptimizer.java

```diff
@@ -242,6 +244,97 @@ private FilterOperator concatFilter(List<Path> fromPaths, FilterOperator operato
     }
   }
 
+  // e.g. translate "select * from root.ln.d1, root.ln.d2 where s1 < 20 AND s2 > 10"
+  // to "select * from root.ln.d1, root.ln.d2 where
+  // (root.ln.d1.s1 < 20 AND root.ln.d1.s2 > 10) OR (root.ln.d2.s1 < 20 AND root.ln.d2.s2 > 10)"
+  private FilterOperator concatFilterByDivice(List<Path> fromPaths, FilterOperator operator)
+          throws LogicalOptimizeException {
+    // remove stars in fromPaths and get deviceId
+    List<String> noStarDevices = removeStarsInDeviceWithUnique(fromPaths);
+
+    FilterOperator filterTwoFolkTree = new FilterOperator(SQLConstant.KW_OR);
+    FilterOperator currentNode = filterTwoFolkTree;
+    FilterOperator parentNode = currentNode;
+    // to check whether duplicate
+    // e.g. SELECT * FROM root.ln.d1, root.ln.d2 where time < 10
+    // brings two 'time < 10' filter operator
+    Set<FilterOperator> operatorSet = new HashSet<>();
+    for (int i = 0; i < noStarDevices.size(); i++) {
+      FilterOperator newOperator = operator.clone();
+      newOperator = concatFilterPath(noStarDevices.get(i), newOperator);
+
+      if (!operatorSet.contains(newOperator)) {
+        if(currentNode.getChildren().size() > 0){
+          FilterOperator newInnerNode = new FilterOperator(SQLConstant.KW_OR);
+          currentNode.addChildOperator(newInnerNode);
+          parentNode = currentNode;
+          currentNode = newInnerNode;
+        }
+        currentNode.addChildOperator(newOperator);
+        operatorSet.add(newOperator);
+      }
+    }
+
+    // if 'OR' has no enough operands due to duplication
+    if(currentNode.getChildren().size() == 1){
+      if(parentNode == currentNode){
+        filterTwoFolkTree = currentNode.getChildren().get(0);
+      } else {
+        parentNode.getChildren().set(1, currentNode.getChildren().get(0));
+      }
+    }
+
+    return filterTwoFolkTree;
+  }
+
+  private List<String> removeStarsInDeviceWithUnique(List<Path> paths)
+          throws LogicalOptimizeException {
+    List<String> retDevices;
+    HashSet<String> deviceSet = new HashSet<>();
+    try {
+      for (Path path : paths) {
+        List<String> all;
+        all = executor.getAllPaths(path.getFullPath());
+        if (all.isEmpty()) {
+          throw new LogicalOptimizeException(
+                  "Path: \"" + path + "\" doesn't correspond to any known time series");
+        }
+        for (String subPath : all) {
+          String deviceId = new Path(subPath).getDevice();
+          if (!deviceSet.contains(deviceId)) {
```

Comment:
I don't think this is a good implementation, say if there are 1000 measurements on average for a device, you will have to perform the meaningless `new Path()` and set operation for 1000 times.
I suggest you write a specific method to match only the device layer.

Reply:
You are right. I didn't find a good function in mManager to remove stars in device level. So I just implement it like this.
This is a new knowledge to me. Maybe someone can help me with that.

Reply:
I have fixed this problem with a new getDevices() function.